source-mysql/replication.go

```diff
@@ -0,0 +1,214 @@
+			logrus.WithField("err", err).Fatal("replication stream error")
+		}
+	}()
+
+	return stream, nil
+}
+
+func splitHostPort(addr string) (string, int64, error) {
+	seps := strings.Split(addr, ":")
+	if len(seps) != 2 {
+		return "", 0, fmt.Errorf("input %q doesn't have <str>:<int> shape", addr)
+	}
+	port, err := strconv.ParseInt(seps[1], 10, 64)
+	if err != nil {
+		return "", 0, fmt.Errorf("invalid port number %q: %w", seps[1], err)
+	}
+	return seps[0], port, nil
+}
+
+type mysqlReplicationStream struct {
+	syncer        *replication.BinlogSyncer
+	streamer      *replication.BinlogStreamer
+	events        chan sqlcapture.ChangeEvent
+	cancel        context.CancelFunc
+	gtidTimestamp time.Time // The OriginalCommitTimestamp value of the last GTID Event
+}
+
+func (rs *mysqlReplicationStream) run(ctx context.Context) error {
+	for {
+		var event, err = rs.streamer.GetEvent(ctx)
+		if err != nil {
+			return fmt.Errorf("error getting next event: %w", err)
+		}
+
+		switch data := event.Event.(type) {
+		case *replication.RowsEvent:
+			var schema, table = string(data.Table.Schema), string(data.Table.Table)
+			if schema == "mysql" {
+				continue
+			}
+
+			var sourceMeta = &mysqlSourceInfo{
+				SourceCommon: sqlcapture.SourceCommon{
+					Millis: rs.gtidTimestamp.UnixMilli(),
+					Schema: schema,
+					Table:  table,
+				},
+			}
+
+			var columnNames = data.Table.ColumnNameString()
+			if len(columnNames) == 0 {
+				return fmt.Errorf("binlog doesn't include column names (are you running with '--binlog-row-metadata=FULL'?)")
+			}
+
+			switch event.Header.EventType {
+			case replication.WRITE_ROWS_EVENTv0, replication.WRITE_ROWS_EVENTv1, replication.WRITE_ROWS_EVENTv2:
+				for _, row := range data.Rows {
+					var after = decodeRow(columnNames, row)
+					rs.events <- sqlcapture.ChangeEvent{
+						Operation: sqlcapture.InsertOp,
+						Source:    sourceMeta,
+						After:     after,
+					}
+				}
+			case replication.UPDATE_ROWS_EVENTv0, replication.UPDATE_ROWS_EVENTv1, replication.UPDATE_ROWS_EVENTv2:
+				for rowIdx := range data.Rows {
+					// Update events contain alternating (before, after) pairs of rows
+					if rowIdx%2 == 1 {
+						var before = decodeRow(columnNames, data.Rows[rowIdx-1])
+						var after = decodeRow(columnNames, data.Rows[rowIdx])
+						rs.events <- sqlcapture.ChangeEvent{
+							Operation: sqlcapture.UpdateOp,
+							Source:    sourceMeta,
+							Before:    before,
+							After:     after,
+						}
+					}
+				}
+			case replication.DELETE_ROWS_EVENTv0, replication.DELETE_ROWS_EVENTv1, replication.DELETE_ROWS_EVENTv2:
+				for _, row := range data.Rows {
+					var before = decodeRow(columnNames, row)
+					rs.events <- sqlcapture.ChangeEvent{
+						Operation: sqlcapture.DeleteOp,
+						Source:    sourceMeta,
+						Before:    before,
+					}
+				}
+			default:
+				return fmt.Errorf("unknown row event type: %q", event.Header.EventType)
```

Comment:
Ya, good idea returning an error here. I was just wondering about this after noticing all the versions above, and I think this is the most reasonable behavior given that a `v3` seems inevitable.

Reply:
Yeah, in general I think it's best to die quickly when we see something unexpected in a replication event stream, since we don't want to make assumptions about the semantics and possibly produce corrupt output. I'm actually torn on having the `v0` and `v1` rows events be accepted, since I haven't personally seen them in my testing.

Reply:
On further reflection I've decided to remove the `v0` row change events from the acceptable list for now. Like I mentioned I've only ever seen the `v2` events in my testing, and I've looked at enough other examples of MySQL binlog replication that I'm comfortable accepting the `v1` events as being identical for our purposes, but I don't have a similar amount of confidence that `v0` change events would work correctly.

Reply:
👍 Sounds good